framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -949,146 +951,175 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
-                }
-                else if (nodeNode instanceof FieldAccessNode) {
-                    // Adapted from FlowExpressionParseUtil.buildFlowExprContextForUse
-
-                    Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory,
-                        ((FieldAccessNode) nodeNode).getReceiver());
+            try {
+                FlowExpressions.Receiver expr = parseExpressionString(expression, flowExprContext,
+                        getCurrentPath(), node, treeForErrorReporting);
 
-                    flowExprContext = new FlowExpressionContext(
-                            internalReceiver, null, checker.getContext());
-                }
-                else if (nodeNode instanceof LocalVariableNode) {
-                    // Adapted from org.checkerframework.dataflow.cfg.CFGBuilder.CFGTranslationPhaseOne.visitVariable
+                CFAbstractStore<?, ?> store = atypeFactory.getStoreBefore(node);
 
-                    ClassTree enclosingClass = TreeUtils
-                            .enclosingClass(getCurrentPath());
-                    TypeElement classElem = TreeUtils
-                            .elementFromDeclaration(enclosingClass);
-                    Node receiver = new ImplicitThisLiteralNode(classElem.asType());
+                CFAbstractValue<?> value = store.getValue(expr);
 
-                    Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory,
-                            receiver);
+                AnnotationMirror inferredAnno = null;
 
-                    flowExprContext = new FlowExpressionContext(
-                            internalReceiver, null, checker.getContext());
+                if (value != null) {
+                    inferredAnno = value.getType().getAnnotationInHierarchy(anno);
                 }
-                else if (nodeNode instanceof ArrayAccessNode) {
-                    // Adapted from FlowExpressionParseUtil.buildFlowExprContextForUse
-
-                    Receiver internalReceiver = FlowExpressions.internalReprOfArrayAccess(atypeFactory,
-                        (ArrayAccessNode) nodeNode);
 
-                    flowExprContext = new FlowExpressionContext(
-                            internalReceiver, null, checker.getContext());
+                if (!checkContract(expr, anno, inferredAnno, store)) {
+                    checker.report(Result.failure(
+                            treeForErrorReporting.getKind() == Tree.Kind.METHOD_INVOCATION
+                                ? "contracts.precondition.not.satisfied"
+                                : "contracts.precondition.not.satisfied.field",
+                            treeForErrorReporting.toString(),
+                            expr == null ? expression : expr.toString()), treeForErrorReporting);
                 }
+            } catch (FlowExpressionParseException e) {
+                // errors are reported at declaration site
```

Comment:
It's not safe. I agree this needs to be fixed, even if it results in the same error being reported multiple times for different lines of code that is being checked (i.e. each time a precondition is checked). However, this is not a simple fix - fixing this causes certain framework and checker tests, as well as Daikon typecheck (for the Nullness Checker), to fail. This is because it exposes the flow expression parser's inability to parse private fields of other classes, and package names. Interestingly, the fix did not make any Lock Checker tests or Daikon typecheck for the Lock Checker fail. I think the Lock Checker should not be gated on this fix. I opened issue #657 to track this.